File: campus-navigate-app/expo-app/src/navigationInput/navigationInputPolicy.ts
import type {
  DerivedNavigationEstimate,
  NavigationInputPolicy,
} from './type';

export const NAVIGATION_INPUT_POLICY: NavigationInputPolicy = {
  maxDerivedUpdatesPerSecond: 15,
  maxRawSamplesInMemory: 32,
  rawSensorRecordingEnabled: false,
  transientRawSensorBatchingEnabled: true,
};

export function assertRawSensorRecordingDisabled(
  policy: NavigationInputPolicy = NAVIGATION_INPUT_POLICY,
) {
  if (policy.rawSensorRecordingEnabled !== false) {
    throw new Error('Raw sensor recording must stay disabled.');
  }
}

export function assertTransientRawSensorBatchingAllowed(
  policy: NavigationInputPolicy = NAVIGATION_INPUT_POLICY,
) {
  if (policy.transientRawSensorBatchingEnabled !== true) {
    throw new Error('Transient raw sensor batching must stay enabled.');
  }
}

export function shouldAcceptDerivedEstimate(input: {
  nextEstimate: DerivedNavigationEstimate;
  previousEstimate: DerivedNavigationEstimate | null;
  policy?: NavigationInputPolicy;
}) {
  const policy = input.policy ?? NAVIGATION_INPUT_POLICY;
  assertRawSensorRecordingDisabled(policy);

  if (!input.previousEstimate) {
    return true;
  }

  const minimumIntervalMs = 1000 / policy.maxDerivedUpdatesPerSecond;
  return (
    input.nextEstimate.timestampMs - input.previousEstimate.timestampMs >=
    minimumIntervalMs
  );
}

File: tarumt-nav-app/flutter_app/lib/domain/navigation_input/logic/navigation_input_policy.dart
import 'package:indoor_navigation/domain/config/app_config.dart';
import 'package:indoor_navigation/domain/navigation/derived_navigation_estimate.dart';
import 'package:indoor_navigation/domain/navigation_input/navigation_input_models.dart';

void assertRawSensorRecordingDisabled([
  NavigationInputPolicy policy = navigationInputPolicy,
]) {
  if (policy.rawSensorRecordingEnabled != false) {
    throw StateError('Raw sensor recording must stay disabled.');
  }
}

void assertTransientRawSensorBatchingAllowed([
  NavigationInputPolicy policy = navigationInputPolicy,
]) {
  if (policy.transientRawSensorBatchingEnabled != true) {
    throw StateError('Transient raw sensor batching must stay enabled.');
  }
}

bool shouldAcceptDerivedEstimate({
  required DerivedNavigationEstimate nextEstimate,
  required DerivedNavigationEstimate? previousEstimate,
  NavigationInputPolicy policy = navigationInputPolicy,
}) {
  assertRawSensorRecordingDisabled(policy);
  if (previousEstimate == null) {
    return true;
  }

  final minimumIntervalMs = 1000 / policy.maxDerivedUpdatesPerSecond;
  return nextEstimate.timestampMs - previousEstimate.timestampMs >=
      minimumIntervalMs;
}
